src/providers/vultr/mod.rs

```diff
@@ -0,0 +1,195 @@
+use crate::providers::MetadataProvider;
+use crate::retry;
+
+#[cfg(test)]
+mod mock_tests;
+
+/// Provider prefix for Vultr Cloud.
+static PROVIDER_PREFIX: &str = "VULTR";
+
+#[derive(Clone, Debug)]
+pub struct VultrProvider {
+    client: retry::Client,
+}
+
+impl VultrProvider {
+    pub fn try_new() -> Result<VultrProvider> {
+        let client = retry::Client::try_new()?.return_on_404(true);
+
+        Ok(VultrProvider { client })
+    }
+
+    #[cfg(test)]
+    fn endpoint_for(name: &str) -> String {
+        let url = mockito::server_url();
+        format!("{}/{}", url, name)
+    }
+
+    #[cfg(not(test))]
+    fn endpoint_for(name: &str) -> String {
+        format!("http://169.254.169.254/v1/{}", name)
+    }
+
+    fn fetch_attribute(
+        &self,
+        map: &mut HashMap<String, String>,
+        key: &str,
+        endpoint: &str,
+    ) -> Result<()> {
+        let content: Option<String> = self
+            .client
+            .get(retry::Raw, Self::endpoint_for(endpoint))
+            .send()?;
+
+        if let Some(value) = content {
+            if !value.is_empty() {
+                map.insert(key.to_string(), value);
+            }
+        }
+
+        Ok(())
+    }
+
+    fn fetch_ssh_keys(&self) -> Result<Vec<String>> {
+        let entries: Option<String> = self
+            .client
+            .get(retry::Raw, VultrProvider::endpoint_for("public-keys"))
+            .send()?;
+        let keys_list = entries.unwrap_or_default();
+
+        let mut keys = Vec::new();
+        for key in keys_list.lines() {
+            let key = key.to_string();
+            keys.push(key)
+        }
+
+        Ok(keys)
+    }
+
+    fn fetch_hostname(&self) -> Result<Option<String>> {
+        let value: Option<String> = self
+            .client
+            .get(retry::Raw, VultrProvider::endpoint_for("hostname"))
+            .send()?;
+
+        let hostname = value.unwrap_or_default();
+        if hostname.is_empty() {
+            return Ok(None);
+        }
+
+        Ok(Some(hostname))
+    }
+}
+
+impl MetadataProvider for VultrProvider {
+    fn attributes(&self) -> Result<HashMap<String, String>> {
+        let mut out = HashMap::with_capacity(20);
+
+        self.fetch_attribute(
+            &mut out,
+            &format!("{}_HOSTNAME", PROVIDER_PREFIX),
+            "hostname",
+        )?;
+
+        self.fetch_attribute(
+            &mut out,
+            &format!("{}_INSTANCE_ID", PROVIDER_PREFIX),
+            "instanceid",
+        )?;
+
+        self.fetch_attribute(
+            &mut out,
+            &format!("{}_REGION_CODE", PROVIDER_PREFIX),
+            "region/regioncode",
+        )?;
+
+        self.fetch_attribute(
+            &mut out,
+            &format!("{}_IPV4_LOCAL", PROVIDER_PREFIX),
+            "bgp/ipv4/my-address",
+        )?;
+
+        self.fetch_attribute(
+            &mut out,
+            &format!("{}_IPV4_PUBLIC", PROVIDER_PREFIX),
+            "interfaces/0/ipv4/address",
```

Comment:
Mmmh, I don't think it works this way.

This is a list of interfaces with their IP details. It seems you arbitrarily picked one as "public" and one as "private", or am I missing something?

Reply:
Thank you so much for the feedback. According to the document https://www.vultr.com/metadata/#v1_interfaces_0_network_type , '0' is public and '1' is private.

Reply:
Perhaps I'm not finding what you are pointing out, but there I see:
> Interface Index - The first interface is labeled with "0", counting onward for subsequent interfaces.

and

> network-type - Retrieve the network type of the specified interface. The value can be one of: "public", "private".